libflux/src/semantic/parser/grammar.md

```diff
@@ -0,0 +1,47 @@
+# Polytype Parser Grammar 
+
+[Modified from Joshua Lapacik's work here](https://github.com/jlapacik/ftel/blob/master/antlr/flux.g4)
+
+```
+'?' = optional; once or none 
+'*' = zero or more times
+'|' = or 
+(values) = a list of possibly repeating values
+single quotes indicate a literal value
+
+
+polytype    = 'forall' '[' vars? ']' ('where' constraints)? monotype
+
+vars        = typeVar (',' typeVar)* 
+constraints = constraint ( (',' | 'and') constraint)* 
+constraint  = typeVar ('is' | ':') kinds
+kinds       = kind ( '+' kind)*
+kind        = IDENTIFIER 
+monotype    = typeVar | typ | array | row | function
+
+type_var    = IDENTIFIER 
+primitives  = INT | FLOAT | STRING | BOOL | DURATION | TIME | REGEX 
+array       = '[' monotype ']'
+row         = '{' properties? '}'
+function    = '(' arguments? ')' '->' monotype
+properties  = property ( '|' property )* ( '|' typeVar)?
+property    = IDENTIFIER ':' monotype
+arguments   = argument ( ',' argument )*
+argument    = required | optional | pipe
+required    = IDENTIFIER ':' monotype
+optional    = '?' IDENTIFIER ':' monotype
+pipe        = '<-' IDENTIFIER? ':' monotype
+
+
+INT         = 'int'
+UINT        = 'uint'
+FLOAT       = 'float'
+STRING      = 'string'
+BOOL        = 'bool'
+DURATION    = 'duration'
+TIME        = 'time'
+REGEX       = 'regular expression'
```

Comment:
Would it be good to have this be `regex` or `regexp` instead? Spaces can mess things up when it comes to scanners and it may make it more clear if this is a single word like the rest of them.

Reply:
I think that in practice I actually went for 'regex' for this. Sorry for the oversight.